README.md

```diff
@@ -80,6 +80,31 @@ If you need a numeric version even for rawhide:
 ```
 
 
+## Caching
+
+To avoid a single point of failure on Bodhi API, you can cache successfully
+fetched data, and use them when there is an issue. For example:
+
+```python
+from fedora_distro_aliases.cache import (
+    save_distro_aliases_cache,
+    load_distro_aliases_cache,
+    BadAliasesCache,
+)
+
+try:
+    try:
+        aliases = get_distro_aliases()
+        save_distro_aliases_cache(aliases)
+    except BodhiDown:
+        aliases = load_distro_aliases_cache()
+    print([x.branch for x in aliases["fedora-all"]])
+
+except BadAliasesCache as ex:
```

Comment:
If I understand it correctly, this gets thrown from `load_distro_aliases_cache()`… wouldn't it be easier to wrap the whole thing for users? Something like:

```py
def get_distro_aliases(with_cache=True, fallback=True):
  …
```

Where `with_cache` triggers the caching, and `fallback` indicates retrieval from cache if Bodhi request fails. Or maybe with different defaults.

Reply:
Hello @mfocko, sorry for the delay here.
The `BadAliasesCache` is thrown by `load_distro_aliases_cache()` when the cache is unusable (the file doesn't exist, is older than allowed, etc). Probably nothing to wrap here since this situation is fatal and the caller can't do anything but print the error to the user.

> wouldn't it be easier to wrap the whole thing for users? Something like:
> Where with_cache triggers the caching, and fallback indicates retrieval from cache if Bodhi request fails. Or maybe with different defaults.

This is where I agree with you. If you take a look at the `README.md` change, there is a proposed usage which leaves a lot of error handling for the caller. This was because of several reasons:

1. @praiskup was concerned about the difficulty of debugging when caching is involved, so I wanted to make it as explicit as possible to the user that caching is indeed happening
2. I wasn't able to figure out how to implement it well
3. It allowed the user to trivially add additional code (such as logging) to the error-handling boilerplate

However, the usage becomes really ugly with the two nested `try` blocks and so on. So, I agree with you, we probably want to wrap the whole thing. Based on your suggestion, we could do this:


```python
def get_distro_aliases(caching=False):
    try:
        releases = bodhi_active_releases()
        if caching:
            save_distro_aliases_cache(releases)
    except requests.exceptions.RequestException as ex:
        if not caching:
            raise BodhiDown from ex
        releases = load_distro_aliases_cache()
    ...
```

But I am considering a bit more generic approach

```python
def get_distro_aliases(cache=None):
    try:
        releases = bodhi_active_releases()
        if cache:
            cache.save(releases)
    except requests.exceptions.RequestException as ex:
        if not cache:
            raise BodhiDown from ex
        releases = cache.load()
    ...


class Cache:
    def __init__(self, path=None, ttl=None):
        self.path = path or PATH
        self.ttl = ttl

    def save(self, releases):
        ...

    def load(self):
        ...

# Usage
cache = Cache(ttl=999)
aliases = get_distro_aliases(cache=cache)
```

This would allow the user to modify the caching parameters like path and how old cache is too old (aka TTL :D). It would also allow the user to provide custom caching implementation, and maybe it could be re-used by opensuse-distro-aliases.

WDYT @mfocko and @dcermak?

Reply:
I like the API proposal from @mfocko for its simplicity.

TBH, I wouldn't make everything user configurable, like the TTL (this is something where we imho can pick a reasonable default).

I am not sure whether I'll really be reusing the caching mechanism, as I don't want to introduce more dependencies to opensuse-distro-aliases. But I might steal your code to avoid the Thursday morning downtime.